Also, the tests are failing:

```
2020-04-29T01:32:59.6316944Z [INFO] Running io.quarkus.vertx.runtime.VertxProducerTest
2020-04-29T01:32:59.7102270Z [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.075 s <<< FAILURE! - in io.quarkus.vertx.runtime.VertxProducerTest
2020-04-29T01:32:59.7105660Z [ERROR] shouldNotFailWithoutConfig  Time elapsed: 0.059 s  <<< ERROR!
2020-04-29T01:32:59.7106503Z java.lang.NullPointerException
```

--- a/extensions/vertx-core/deployment/src/main/java/io/quarkus/vertx/core/deployment/VertxCoreProcessor.java
+++ b/extensions/vertx-core/deployment/src/main/java/io/quarkus/vertx/core/deployment/VertxCoreProcessor.java
@@ -1,5 +1,9 @@
 package io.quarkus.vertx.core.deployment;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Consumer;
 import java.util.function.Supplier;
 
 import javax.inject.Singleton;
@@ -26,6 +30,7 @@ import io.quarkus.vertx.core.runtime.VertxLogDelegateFactory;
 import io.quarkus.vertx.core.runtime.config.VertxConfiguration;
 import io.vertx.core.AbstractVerticle;
 import io.vertx.core.Vertx;
+import io.vertx.core.VertxOptions;
 import io.vertx.core.spi.resolver.ResolverProvider;
 
 class VertxCoreProcessor {
@@ -66,11 +71,18 @@ class VertxCoreProcessor {
     @Record(value = ExecutionTime.RUNTIME_INIT)
     CoreVertxBuildItem build(VertxCoreRecorder recorder,
             LaunchModeBuildItem launchMode, ShutdownContextBuildItem shutdown, VertxConfiguration config,
+            List<VertxOptionsConsumerBuildItem> vertxOptionsConsumers,
             BuildProducer<SyntheticBeanBuildItem> syntheticBeanBuildItemBuildProducer,
             BuildProducer<ServiceStartBuildItem> serviceStartBuildItem) {
 
+        Collections.sort(vertxOptionsConsumers);
+        List<Consumer<VertxOptions>> consumers = new ArrayList<>(vertxOptionsConsumers.size());
+        for (VertxOptionsConsumerBuildItem x : vertxOptionsConsumers) {
+            consumers.add(x.getConsumer());
+        }
+
         Supplier<Vertx> vertx = recorder.configureVertx(config,
-                launchMode.getLaunchMode(), shutdown);
+                launchMode.getLaunchMode(), shutdown, consumers);
         syntheticBeanBuildItemBuildProducer.produce(SyntheticBeanBuildItem.configure(Vertx.class)
                 .types(Vertx.class)
                 .scope(Singleton.class)

--- a/extensions/vertx-core/deployment/src/main/java/io/quarkus/vertx/core/deployment/VertxOptionsConsumerBuildItem.java
+++ b/extensions/vertx-core/deployment/src/main/java/io/quarkus/vertx/core/deployment/VertxOptionsConsumerBuildItem.java
@@ -1,0 +1,33 @@
+package io.quarkus.vertx.core.deployment;
+
+import java.util.function.Consumer;
+
+import io.quarkus.builder.item.MultiBuildItem;
+import io.vertx.core.VertxOptions;
+
+/**
+ * Provide a consumer of VertxOptions to allow customization of
+ * Vert.x system behavior, e.g. setting MetricsOptions to enable
+ * and configure a metrics provider.
+ * <p>
+ * Consumers will be called in priority order (lowest to highest)
+ * after VertxConfiguration has been read and applied.
+ */
+public final class VertxOptionsConsumerBuildItem extends MultiBuildItem implements Comparable<VertxOptionsConsumerBuildItem> {
+    private final Consumer<VertxOptions> optionsConsumer;
+    private final int priority;
+
+    public VertxOptionsConsumerBuildItem(Consumer<VertxOptions> optionsConsumer, int priority) {
+        this.optionsConsumer = optionsConsumer;
+        this.priority = priority;
+    }
+
+    public Consumer<VertxOptions> getConsumer() {
+        return optionsConsumer;
+    }
+
+    @Override
+    public int compareTo(VertxOptionsConsumerBuildItem o) {
+        return Integer.compare(this.priority, o.priority);
+    }
+}

--- a/extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java
+++ b/extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java
@@ -11,6 +11,7 @@ import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
 import java.util.function.Supplier;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
@@ -49,8 +50,8 @@ public class VertxCoreRecorder {
     static volatile VertxSupplier vertx;
 
     public Supplier<Vertx> configureVertx(VertxConfiguration config,
-            LaunchMode launchMode, ShutdownContext shutdown) {
-        vertx = new VertxSupplier(config);
+            LaunchMode launchMode, ShutdownContext shutdown, List<Consumer<VertxOptions>> customizers) {
+        vertx = new VertxSupplier(config, customizers);
         if (launchMode != LaunchMode.DEVELOPMENT) {
             // we need this to be part of the last shutdown tasks because closing it early (basically before Arc)
             // could cause problem to beans that rely on Vert.x and contain shutdown tasks
@@ -77,12 +78,18 @@ public class VertxCoreRecorder {
         return vertx;
     }
 
-    public static Vertx initialize(VertxConfiguration conf) {
-        if (conf == null) {
-            return logVertxInitialization(Vertx.vertx());
+    public static Vertx initialize(VertxConfiguration conf, VertxOptionsCustomizer customizer) {
+
+        VertxOptions options = new VertxOptions();
+
+        if (conf != null) {
+            convertToVertxOptions(conf, options, true);
         }
 
-        VertxOptions options = convertToVertxOptions(conf, true);
+        // Allow extension customizers to do their thing
+        if (customizer != null) {
+            customizer.customize(options);
+        }
 
         Vertx vertx;
         if (options.getEventBusOptions().isClustered()) {
@@ -106,12 +113,11 @@ public class VertxCoreRecorder {
         return vertx;
     }
 
-    private static VertxOptions convertToVertxOptions(VertxConfiguration conf, boolean allowClustering) {
+    private static VertxOptions convertToVertxOptions(VertxConfiguration conf, VertxOptions options, boolean allowClustering) {
+
         if (!conf.useAsyncDNS) {
             System.setProperty(ResolverProvider.DISABLE_DNS_RESOLVER_PROP_NAME, "true");
         }
-
-        VertxOptions options = new VertxOptions();
 
         if (allowClustering) {
             // Order matters, as the cluster options modifies the event bus options.
@@ -318,18 +324,35 @@ public class VertxCoreRecorder {
 
     static class VertxSupplier implements Supplier<Vertx> {
         final VertxConfiguration config;
+        final VertxOptionsCustomizer customizer;
         Vertx v;
 
-        VertxSupplier(VertxConfiguration config) {
+        VertxSupplier(VertxConfiguration config, List<Consumer<VertxOptions>> customizers) {
             this.config = config;
+            this.customizer = new VertxOptionsCustomizer(customizers);
         }
 
         @Override
         public synchronized Vertx get() {
             if (v == null) {
-                v = initialize(config);
+                v = initialize(config, customizer);
             }
             return v;
+        }
+    }
+
+    static class VertxOptionsCustomizer {
+        final List<Consumer<VertxOptions>> customizers;
+
+        VertxOptionsCustomizer(List<Consumer<VertxOptions>> customizers) {
+            this.customizers = customizers;
+        }
+
+        VertxOptions customize(VertxOptions options) {
+            for (Consumer<VertxOptions> x : customizers) {
+                x.accept(options);
+            }
+            return options;
         }
     }
 }

--- a/extensions/vertx-core/runtime/src/test/java/io/quarkus/vertx/core/runtime/VertxCoreProducerTest.java
+++ b/extensions/vertx-core/runtime/src/test/java/io/quarkus/vertx/core/runtime/VertxCoreProducerTest.java
@@ -1,16 +1,18 @@
 package io.quarkus.vertx.core.runtime;
 
-import static org.junit.jupiter.api.Assertions.assertTrue;
-import static org.junit.jupiter.api.Assertions.fail;
-
 import java.time.Duration;
+import java.util.Arrays;
 import java.util.Optional;
 import java.util.OptionalInt;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
 
 import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
+import io.quarkus.vertx.core.runtime.VertxCoreRecorder.VertxOptionsCustomizer;
 import io.quarkus.vertx.core.runtime.config.ClusterConfiguration;
 import io.quarkus.vertx.core.runtime.config.EventBusConfiguration;
 import io.quarkus.vertx.core.runtime.config.JksConfiguration;
@@ -18,6 +20,8 @@ import io.quarkus.vertx.core.runtime.config.PemKeyCertConfiguration;
 import io.quarkus.vertx.core.runtime.config.PemTrustCertConfiguration;
 import io.quarkus.vertx.core.runtime.config.PfxConfiguration;
 import io.quarkus.vertx.core.runtime.config.VertxConfiguration;
+import io.vertx.core.Vertx;
+import io.vertx.core.VertxOptions;
 
 public class VertxCoreProducerTest {
 
@@ -49,11 +53,26 @@ public class VertxCoreProducerTest {
         configuration.cluster = cc;
 
         try {
-            VertxCoreRecorder.initialize(configuration);
-            fail("It should not have a cluster manager on the classpath, and so fail the creation");
+            VertxCoreRecorder.initialize(configuration, null);
+            Assertions.fail("It should not have a cluster manager on the classpath, and so fail the creation");
         } catch (IllegalStateException e) {
-            assertTrue(e.getMessage().contains("No ClusterManagerFactory"));
+            Assertions.assertTrue(e.getMessage().contains("No ClusterManagerFactory"),
+                    "The message should contain ''. Message: " + e.getMessage());
         }
+    }
+
+    @Test
+    public void shouldInvokeCustomizers() {
+        final AtomicBoolean called = new AtomicBoolean(false);
+        VertxOptionsCustomizer customizers = new VertxOptionsCustomizer(Arrays.asList(
+                new Consumer<VertxOptions>() {
+                    @Override
+                    public void accept(VertxOptions vertxOptions) {
+                        called.set(true);
+                    }
+                }));
+        Vertx v = VertxCoreRecorder.initialize(createDefaultConfiguration(), customizers);
+        Assertions.assertTrue(called.get(), "Customizer should get called during initialization");
     }
 
     private VertxConfiguration createDefaultConfiguration() {

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
@@ -138,7 +138,7 @@ public class VertxHttpRecorder {
         }
         VertxConfiguration vertxConfiguration = new VertxConfiguration();
         ConfigInstantiator.handleObject(vertxConfiguration);
-        Vertx vertx = VertxCoreRecorder.initialize(vertxConfiguration);
+        Vertx vertx = VertxCoreRecorder.initialize(vertxConfiguration, null);
 
         try {
             HttpConfiguration config = new HttpConfiguration();

--- a/extensions/vertx/runtime/src/test/java/io/quarkus/vertx/runtime/VertxProducerTest.java
+++ b/extensions/vertx/runtime/src/test/java/io/quarkus/vertx/runtime/VertxProducerTest.java
@@ -27,7 +27,7 @@ public class VertxProducerTest {
 
     @Test
     public void shouldNotFailWithoutConfig() {
-        producer.vertx = VertxCoreRecorder.initialize(null);
+        producer.vertx = VertxCoreRecorder.initialize(null, null);
         verifyProducer();
     }
 

